### ISSUE TITLE
What is the purpose of escape nodes?

### ISSUE THREAD

**Issue by eush77** [wooorm/remark#71]
I am wondering why escape nodes are in the AST? Are they ever useful?

Wouldn't it be simpler for users if

```
_foo\_ bar_
```

was parsed as

```json
{                                                                                   
  "type": "paragraph",                                                              
  "children": [                                                                     
    {                                                                               
      "type": "emphasis",                                                           
      "children": [                                                                 
        {                                                                           
          "type": "text",                                                           
          "value": "foo_ bar"                                                       
        }                                                                           
      ]                                                                             
    }                                                                               
  ]                                                                                 
}
```

and not as

```json
{
  "type": "paragraph",
  "children": [
    {
      "type": "emphasis",
      "children": [
        {
          "type": "text",
          "value": "foo"
        },
        {
          "type": "escape",
          "value": "_"
        },
        {
          "type": "text",
          "value": " bar"
        }
      ]
    }
  ]
}
```

It seems to me that escaping information is aimed primarily and solely at parser and plugins shouldn't even care. Commonmark spec [says](http://spec.commonmark.org/0.22/#backslash-escapes):

> Escaped characters are treated as regular characters and do not have their usual Markdown meanings.

In my interpretation, this means that both `_foo\_bar_baz_`, `_foo\_bar\_baz_`, and `_foo_bar_baz_` should result in the exact same syntax tree since in both cases inner underscores have no associated semantic meaning.

There is one use case for escape nodes that is special, I found, and that is using escape nodes with `value="\n"` as [hard line breaks](http://spec.commonmark.org/0.22/#hard-line-breaks). [Example](https://github.com/wooorm/mdast-html/blob/4315c6e95cf64027890ea4da79ec71ab76071eca/lib/compilers.js#L897-L899) from [mdast-html](https://github.com/wooorm/mdast-html):

```js
function escape(node) {
    return this[node.value === '\n' ? 'break' : 'text'](node);
}
```

But this is strange because mdast already has a break node. For example,

```
foo  
bar
```

(with two spaces at end of the first line) renders into

```json
{
  "type": "paragraph",
  "children": [
    {
      "type": "text",
      "value": "foo"
    },
    {
      "type": "break"
    },
    {
      "type": "text",
      "value": "bar"
    }
  ]
}
```

However,

```
foo\
bar
```

renders into

```json
{
  "type": "paragraph",
  "children": [
    {
      "type": "text",
      "value": "foo"
    },
    {
      "type": "escape",
      "value": "\n"
    },
    {
      "type": "text",
      "value": "bar"
    }
  ]
}
```

[According to](http://spec.commonmark.org/0.22/#hard-line-breaks) Commonmark both cases should be equivalent:

> For a more visible alternative, a backslash before the line ending may be used instead of two spaces.

So why not create break nodes in both cases and forego escape nodes?

My perspective on this is limited and maybe I'm getting it all wrong. Could you correct me then and point at the right direction?

**Comment by eush77** [comment]
Related: #26.

**Comment by wooorm** [comment]
Thanks @eush77 for the extensive write-down!

I agree on the break-nodes in CommonMark being arguably wrongly classified, at least, [they should change](https://github.com/wooorm/mdast/issues/26).

Regarding why escape-nodes even exist: I started out without them, but found that it’s really hard to get them back properly. The problem is, they are not needed that often, but when they are, it’s hard for a computer to figure out. The document author probably knows it a lot better.

As an example of the characters which need to be escaped, check out the [list in **mdast-util-escape**](https://github.com/wooorm/mdast-util-escape/blob/391bb841bfe4e86fe66ebc071ae49273031cde28/index.js#L21-L36): most of them characters which only have a special meaning in select cases.

Some of them can probably be figured out by the compiler but it’s not very easy. For example, the `tick` can easily, by accident, produce inline code spanning over multiple lines.

**Comment by eush77** [comment]
After parsing `_foo_bar\_baz_` and constructing the AST, what difference does it make knowing that the second underscore was escaped and the first was not? Could you provide an example?

In the classical compiler architecture, escape characters are handled in the scanner that produces a sequence of tokens, and it is the scanner (not even parser) on which level the notion of character escaping is usually abstracted away.

Take a look at JavaScript, for example. The language uses parentheses and escape sequences in the language, but there're no such things as ParenthesizedExpression or EscapeLiteral in [estree](https://github.com/estree/estree) spec. It makes me wonder what makes mdast different.

Maybe the approach taken by acorn and esprima is worth considering. They parse `"foo\_bar"` as

```json
{
  "type": "Literal",
  "value": "foo_bar",
  "raw": "\"foo\\_bar\""
}
```

Adding `raw` attribute to `text` nodes is arguably better than escaping in that users can ignore it if they don't need such information. Does it solve the problems you need escape nodes for?

**Comment by wooorm** [comment]
Okay, I’ll try removing escape nodes on a branch, but my difficulties with this stem from the fact that the compiler needs significant more logic to correctly escape raw characters where needed (escaping all escapable characters adds *a lot* of slashes).

For example, the `>` is probably pretty simple: only escape it at the start of a line; the parens (`()`), square brackets (`[]`), ticks (`` ` ``) are a lot harder.

**Comment by eush77** [comment]
Why would you escape *all* escapable characters? If you go with `raw` property approach, this property holds the lexeme in exactly the same form in which it appears in the source string so that `value` could be *completely unescaped*. E.g. AST for `_a_b\_c_` would have a text node with `raw="a_b\\_c"` and `value="a_b_c"`. The first underscore is not escaped in `raw` because it wasn't originally escaped.

If you have things you are more interested in than removing escape nodes, I could try to help you with that.

**Comment by eush77** [comment]
Also, if there's a technical limitation to it, I would even imagine doing the transformation in a postprocessing step.

Maybe as simple as:

```js
visit(ast, 'escape', function (escape, index, parent) {
  parent.children[index] = {
    type: 'text',
    value: escape.value,
    raw: '\\' + escape.value
  };
 });

visit(ast, 'text', function (text) {
  text.raw = text.value;
});
```

followed by merging adjacent text nodes to reduce the node count.

**Comment by wooorm** [comment]
@eush77 `raw` properties would be nice to have, but the compiler also has to decide which escapable characters to escape on nodes *without* that `raw` property!

**Comment by eush77** [comment]
One possible issue with `raw` properties is that they are not so easy to update.

```js
mdast.use(function () {
  return function (ast) {
    var text = ast.children[0].children[0];
    text.value = text.value.replace(/;/g, '+');
    // text.raw = ...
  };
}).process('a&amp;=b; &notanentity;');
```

**Comment by eush77** [comment]
> `raw` properties would be nice to have, but the compiler also has to decide which escapable characters to escape on nodes without that raw property!

Okay, but the compiler has to do it anyway:

```js
mdast.use(function () {
  return function (ast) {
    ast.children[0].children[0].children[0].value = 'a_ b';
  };
}).process('_a_')
//=> '_a_ b_\n' 
// Should probably be "_a\\_ b_".
```

**Comment by wooorm** [comment]
Right, so I’m not sure if `raw` properties are really needed; the compiler cannot rely on them, plug-ins won’t keep them up-to-date*, it makes the last more verbose, and they can already be gathered by checking `file.toString().slice(node.position.start.offset, node.position.end.offset)` when using `mdast-range`.

\* I even think this should be discouraged; similar to how `position`s on generated nodes should be void.

**Comment by wooorm** [comment]
@eush77 I just remembered you saying you [might be able to help](https://github.com/wooorm/mdast/issues/71#issuecomment-143495666), if that’s still the case (I’m quite busy IRL, with entities, and with all those other issues), I’d love to take you up on that :+1:

I’d suggest first trying without `raw` (maybe add another issue for that?), then, replace the spot where `.type = 'escape'` is set with `'text'` (also change this in the tests).

Then, try to get the tests too pass; they check if a document can be compiled to an AST (A), which is compiled, which is parsed to another AST (B), and check if A and B are equal.

If all those tests pass, I’m fine with merging.

If they don’t, we could also try to move most escapes to text (e.g., `+`), while still keeping some as escapes (e.g., `` ` ``), but that’s not perfect!

**Comment by eush77** [comment]
@wooorm I would be happy to help!

**Comment by eush77** [comment]
@wooorm Since in `gfm` and `commonmark` modes different [characters](https://github.com/wooorm/mdast/blob/b28a45ac3d8291ef2fb3e173468ac992f8c56f1f/lib/expressions.js#L34) can be backslash-escaped, escaping them in the compiler would require adding `gfm` and `commonmark` options to `mdast.stringify`, but being able to parse document in just GFM and stringify it in Commonmark doesn't seem useful. Is this the way to go?

**Comment by wooorm** [comment]
@eush77 I think so; most people will use `process`, which already passes all `options` to both `parse` and `compile`. It should still be in the docs, but I do agree that this is the way to go!

**Comment by wooorm** [comment]
This will be closed by GH-78 in 3.0.0.
